pkg/goldpinger/k8s.go

```diff
@@ -41,10 +43,68 @@ func getPodNamespace() string {
 	return namespace
 }
 
+func getAllNodes() *v1.NodeList {
+	timer := GetLabeledKubernetesCallsTimer()
+	nodes, err := GoldpingerConfig.KubernetesClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		zap.L().Error("Error getting nodes", zap.Error(err))
+		CountError("kubernetes_api")
+	} else {
+		timer.ObserveDuration()
+	}
+	return nodes
+}
+
+// getHostIP gets the IP of the host where the pod is scheduled. If UseIPv6 is enabled then we need to check
+// the node IPs since HostIP will only list the default IP version one.
+func getHostIP(p v1.Pod) string {
+	if !GoldpingerConfig.UseIPv6 || k8snet.IsIPv6String(p.Status.HostIP) {
+		return p.Status.HostIP
+	}
+
+	nodes := getAllNodes()
+
+	for _, node := range nodes.Items {
+		var v6Addr string
+		var isNodeForPod bool
+		for _, addr := range node.Status.Addresses {
+			if addr.Address == p.Status.HostIP {
+				isNodeForPod = true
+			}
+
+			if k8snet.IsIPv6String(addr.Address) {
+				v6Addr = addr.Address
+			}
+		}
+
+		if isNodeForPod && len(v6Addr) > 0 {
+			return v6Addr
+		}
+	}
+	return p.Status.HostIP
+}
+
+// getPodIP will get an IPv6 IP from PodIPs if the UseIPv6 config is set, otherwise just return the object PodIP
+func getPodIP(p v1.Pod) string {
+	if !GoldpingerConfig.UseIPv6 {
+		return p.Status.PodIP
+	}
+
+	var v6IP string
+	if p.Status.PodIPs != nil {
+		for _, ip := range p.Status.PodIPs {
+			if k8snet.IsIPv6String(ip.IP) {
+				v6IP = ip.IP
```

Comment:
so we'd never ping both. To test dual stack would we have two deployments one ipv4 and one ipv6?

Reply:
pinging both might be more robust but would require more work. it seems feasible to have separate deployments for each ip version for now. In a pipeline you could do `create ipv4 daemonset -> run -> collect metrics -> teardown -> create ipv6 daemonset -> run -> collect metrics -> teardown`